CSP-1776: Error in both VacancyApprovedEvent & CohortAssignedToProviderEvent

shaun-downey-education opened the pull request:
- Updated **VacancyApprovedEventHandler** & **CohortAssignedToProviderEventHandler** to record job audits with message details.

- Remove **AsNoTracking()** from **AccountLegalEntity**, **AccountProvider**, & **Provider** repositories.

- Updated insert methods on inserting of **AccountProviderLegalEntities** to use tracked entities.

DevenKShah committed:
CSP-1776 Added commitments mock server

shaun-downey-education committed:
CSP-1776: Updated VacancyApprovedEventHandler & CohortAssignedToProviderEventHandler to record job audits with message details.

Remove AsNoTracking() from AccountLegalEntity, AccountProvider, & Provider repositories.

Updated insert methods on inserting of AccountProviderLegalEntities to use tracked entities.

shaun-downey-education committed:
CSP-1776: Sonarcloud code suggestion, additional unit test coverage.

shaun-downey-education committed:
Merge branch 'main' into CSP-1776-ErrorInCohortsEvent

DevenKShah commented on src/SFA.DAS.PR.Jobs/MessageHandlers/Recruit/VacancyApprovedEventHandler.cs:
```diff
@@ -90,26 +91,43 @@ await _accountProviderRepository.AddAccountProvider(
+        );
 
-        Notification notification = CreateNotification("LinkedAccountRecruit", "PR Jobs: VacancyReviewedEvent", accountLegalEntity.Id, liveVacancy);
+        await _providerRelationshipsDataContext.SaveChangesAsync(context.CancellationToken);
```
Remove this save changes call

DevenKShah commented on src/SFA.DAS.PR.Jobs/MessageHandlers/CohortAssignedToProviderEventHandler.cs:
```diff
@@ -67,7 +72,7 @@ public async Task Handle(CohortAssignedToProviderEvent message, IMessageHandlerC
+            accountProvider = CreateAccountProvider(cohort.AccountId, cohort.ProviderId);
         }
 
         AccountProviderLegalEntity? accountProviderLegalEntity = await _accountProviderLegalEntityRepository.GetAccountProviderLegalEntity(
```
If account provider above is null, this call to get account provider legal entity is redundant, we can assume its absence.

DevenKShah commented on src/SFA.DAS.PR.Jobs/MessageHandlers/Recruit/VacancyApprovedEventHandler.cs:
```diff
@@ -90,26 +91,43 @@ await _accountProviderRepository.AddAccountProvider(
 
-        await _providerRelationshipsDataContext.PermissionsAudit.AddAsync(permissionAudit, context.CancellationToken);
+        _providerRelationshipsDataContext.Notifications.Add(
+            CreateNotification("LinkedAccountRecruit", "PR Jobs: VacancyReviewedEvent", accountLegalEntity.Id, liveVacancy)
```
Use `System` in CreatedBy

DevenKShah commented on src/SFA.DAS.PR.Jobs/MessageHandlers/CohortAssignedToProviderEventHandler.cs:
```diff
@@ -137,17 +159,16 @@ await _permissionAuditRepository.CreatePermissionAudit(
             CreatedDate = DateTime.UtcNow,
             Ukprn = ukprn,
             AccountLegalEntityId = accountLegalEntityId,
             CreatedBy = "PR Jobs: CohortAssignedToProviderEvent",
```
Use `System` in CreatedBy

shaun-downey-education committed:
CSP-1776: Remove check for AccountProviderLegalEntity when AccountProvider is null on the CohortAssignedToProviderEventHandler.

Updated the CohortAssignedToProviderEventHandler & VacancyApprovedEventHandler to record 'System' against the Notification CreatedBy record.

shaun-downey-education committed:
CSP-1776: Marked CreateLiveVacanyModel as static.

shaun-downey-education committed:
CSP-1776: Refactor CohortAssignedToProviderEvent & VacancyApprovedEvent logic into a service.

DevenKShah commented on src/SFA.DAS.PR.Jobs/Services/RelationshipService.cs:
```diff
@@ -0,0 +1,168 @@
+
+public interface IRelationshipService
+{
+    Task CreateRelationship<T>(ILogger<T> _logger, RelationshipModel relationshipModel, CancellationToken cancellationToken) where T : class;
```
No need to take logger as parameter, can use injection

DevenKShah commented on src/SFA.DAS.PR.Jobs/Services/RelationshipService.cs:
```diff
@@ -0,0 +1,168 @@
+{
+    private const string SystemUser = "System";
+
+    public async Task CreateRelationship<T>(ILogger<T> _logger, RelationshipModel relationshipModel, CancellationToken cancellationToken) where T : class
```
Instead of using model with nullable properties, can we have two overloads, it will require breaking down this method in to smaller private units which each overload will invoke as needed?

shaun-downey-education committed:
CSP-1776: Remove logger as a parameter from the relationship service.

MarkFCain commented on src/SFA.DAS.PR.Data/Repositories/AccountLegalEntityRepository.cs:
```diff
@@ -21,7 +21,6 @@ public class AccountLegalEntityRepository(IProviderRelationshipsDataContext _pro
     public async Task<AccountLegalEntity?> GetAccountLegalEntity(string publicHashedId, CancellationToken cancellationToken)
     {
         return await _providerRelationshipsDataContext.AccountLegalEntities
-            .AsNoTracking()
```
is there a reason this is taken out?

MarkFCain commented on src/SFA.DAS.PR.Data/Repositories/AccountProviderRepository.cs:
```diff
@@ -14,7 +14,6 @@ public class AccountProviderRepository(IProviderRelationshipsDataContext _provid
     public async ValueTask<AccountProvider?> GetAccountProvider(long providerUkprn, long accountId, CancellationToken cancellationToken)
     {
         return await _providerRelationshipsDataContext.AccountProviders
-            .AsNoTracking()
```
is there a reason this is taken out?

DevenKShah committed:
CSP-1776 Fixed recruit notification not persisting legal entity id

DevenKShah committed:
CSP-1776 Dispose context in unit tests
